DM-41158: (mostly) implement QueryDriver for DirectButler

TallJimbo opened the pull request:
I've inspected all of the missed-coverage lines and they fall into two categories:
 - they need `DatasetRef` or `DataCoordinate` result support to be tested (which I'm deferring to another ticket);
 - they're really trivial error cases that I don't really think are worth the effort.

## Checklist

- [x] ran Jenkins
- [ ] ~added a release note for user-visible changes to `doc/changes`~ (N/A)

TallJimbo committed:
Implement QueryDriver for DirectButler.

TallJimbo committed:
Move NameShrinker out of registry and add container methods.

TallJimbo committed:
Accept NameShrinker in ColumnSpec.to_sql_spec.

TallJimbo committed:
Guard against long field names in dimensions manager.

It is very unlikely that we'll ever trip this guard.

TallJimbo committed:
Default QueryTree argument to Query constructor.

This reduces the need to import more stuff from the queries subpackage
in butlers.

TallJimbo committed:
Implement Butler._query.

TallJimbo committed:
Move 'dimensions' attribute out of QueryBase/Query.

Query already has 'constraint_dimensions' (the full dimensions that
can be used for e.g. WHERE expressions), which is different from
what 'dimensions' should have meant (the dimensions of the output
objects) on the QueryResults classes.

TallJimbo committed:
Fix truncated docstring in queries.tree.ColumnSet.

TallJimbo committed:
Make queries.ExpressionFactory friendlier to type checkers.

Previous annotations rejected a lot of typical usage that would almost
always be fine (because most DimensionElements are Dimensions and are
hence usable directly as values).

TallJimbo committed:
Fix doc typo that inverted meaning in Database interface.

TallJimbo committed:
Guard against skypix regions in result specs.

SkyPix regions should always be computed on the fly in the client.

TallJimbo committed:
Export InvalidQueryError at package scope.

TallJimbo committed:
Fix typo in comment.

TallJimbo committed:
Don't lift 'queries' symbols into package scope (yet).

They're not public, and some of them clash with Registry symbols.

TallJimbo committed:
Fix incorrect query-result type imports.

This test wanted the result objects from

   lsst.daf.butler.registry.queries

for use in annotations, but it was getting them from

   lsst.daf.butler.queries

and we didn't notice because we don't run MyPy on tests.

dhirving commented on python/lsst/daf/butler/column_spec.py:
```diff
@@ -79,11 +80,14 @@ class _BaseColumnSpec(pydantic.BaseModel, ABC):
     )
 
-    def to_sql_spec(self, **kwargs: Any) -> ddl.FieldSpec:
+    def to_sql_spec(self, name_shrinker: NameShrinker | None = None, **kwargs: Any) -> ddl.FieldSpec:
```
Should this nameshrinker argument be non-optional?  I would think that if it is ever needed, it will always be needed... otherwise different parts of the code can think names are different.  Just creates the potential for certain columns to work with some types of queries/operations but not others.

TallJimbo replied:
I prefer to do that at a higher level, because the set of names that need shrinking is limited to only those constructed from dataset type names, and that keeps name-shrinker arguments from sprouting up in many more places.

dhirving commented on python/lsst/daf/butler/direct_query_driver/_driver.py:
```diff
@@ -0,0 +1,1200 @@
+        # complicated to implement anyway.
+        #
+        # We start a transaction rather than just opening a connection both
+        # to establish consistency guarantees throughout the query context and
```
What consistency guarantees do we get from the transaction?  I thought Butler transactions were always READ COMMITTED, which is the same as what we'd get without the explicit transaction.

TallJimbo replied:
Yeah, there was some lazy thinking going on here; I had been mostly thinking about consistency of things in caches (which you reminded me I hadn't actually done in another thread) and started conflating that with actual transaction consistency (which as you note here depends on isolation level).  I'll drop the first clause of this sentence and just keep the second.

dhirving commented on tests/test_query_direct_sqlite.py:
```diff
@@ -0,0 +1,65 @@
+TESTDIR = os.path.abspath(os.path.dirname(__file__))
+
+
+class DirectButlerSQLiteTests(ButlerQueryTests, unittest.TestCase):
```
Should there be another instance of these tests for Postgres as well?

TallJimbo replied:
Yes, I meant to do that before I put it up for review and apparently I forgot.

But having added it now, I see some failures in cursor/temp-table lifetimes.  I wasn't able to figure it quickly, so I've marked the tests that use temp tables as expected failures and will try to fix them on [DM-43697](https://rubinobs.atlassian.net/browse/DM-43697).

dhirving commented on python/lsst/daf/butler/direct_query_driver/_driver.py:
```diff
@@ -0,0 +1,1200 @@
+        self._universe = universe
+        self._defaults = defaults
+        self._materializations: dict[
+            qt.MaterializationKey, tuple[sqlalchemy.Table, frozenset[str], Postprocessing]
```
Making this tuple a NamedTuple or frozen dataclass would make it self-documenting and less fiddly to interact with.

dhirving commented on python/lsst/daf/butler/direct_query_driver/_driver.py:
```diff
@@ -0,0 +1,1200 @@
+        assert self._exit_stack is not None
+        while self._cursors:
+            _, cursor = self._cursors.popitem()
+            cursor.close(exc_type, exc_value, traceback)
```
It might be a good idea to push this cursor closing onto the exit stack instead of directly calling close... that way the rest of cleanup will still occur even if closing a cursor raises an exception.

TallJimbo replied:
I tried something like that first, but got stuck being unable to pop a cursor out of the exit stack by name when it could be closed due to normal termination of iteration.  But you're right that there's a resource management gap here.  Adding a manual try/except here is nontrivial because we do want to re-raise at the end, so I'll punt this to [DM-43697](https://rubinobs.atlassian.net/browse/DM-43697).

dhirving replied:
I was just thinking call `exit_stack.push(cursor)` right here instead of calling `cursor.close()`... `push()` calls `__exit__` without calling `__enter__`.

dhirving commented on python/lsst/daf/butler/direct_query_driver/_driver.py:
```diff
@@ -0,0 +1,1200 @@
+            Traceback as obtained from `sys.exc_info`, or `None` if there was
+            no error.
+        """
+        self._context.__exit__(exc_type, exc_value, traceback)
```
Is it allowed to call `__exit__` on a SQLAlchemy cursor multiple times?  If not it might be a good idea to guard against that here.

TallJimbo replied:
It's not _quite_ as clearly stated as I'd like, but [the docs](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.CursorResult.close) say

> This closes out the underlying DBAPI cursor corresponding to the statement execution, if one is still present.

dhirving commented on python/lsst/daf/butler/direct_query_driver/_driver.py:
```diff
@@ -0,0 +1,1200 @@
+        result_page = cursor.next()
+        if result_page.next_key is None:
+            # Cursor has been exhausted; close it here.
+            cursor.close()
```
This cursor closing logic is already handled inside `cursor.next()`.

dhirving commented on python/lsst/daf/butler/direct_query_driver/_driver.py:
```diff
@@ -0,0 +1,1200 @@
+                    m = raw_row._mapping
+                    d = {k: m[v] for k, v in column_map}
+                    d["timespan"] = self._timespan_repr_cls.extract(m, name=timespan_qualified_name)
+                    record_set.add(record_cls(**d))
```
It might be better not to duplicate the logic for the non-timespan columns between these two branches... creates the potential for them to get out of sync as things change.
```suggestion
            if result_spec.element.temporal:
                timespan_qualified_name = qt.ColumnSet.get_qualified_name(
                    result_spec.element.name, "timespan"
                )
            else:
                timespan_qualified_name = None
                
            for raw_row in raw_rows:
                m = raw_row._mapping
                d = {k: m[v] for k, v in column_map}
                if timespan_qualified_name is not None:
                    d["timespan"] = self._timespan_repr_cls.extract(m, name=timespan_qualified_name)
                record_set.add(record_cls(**d))
```

TallJimbo replied:
Yeah, this was intentional - but possibly premature - optimization; I wanted to bring as much out of the loop over result rows as possible. But your version is all that different in that respect.

dhirving commented on python/lsst/daf/butler/direct_query_driver/_driver.py:
```diff
@@ -0,0 +1,1200 @@
+                "QueryDriver context must be entered before 'upload_data_coordinates' is called."
+            )
+        columns = qt.ColumnSet(dimensions).drop_implied_dimension_keys()
+        table_spec = ddl.TableSpec(
```
For maintainability it might be a good idea to deduplicate this ColumnSet-to-TableSpec logic between here and `QueryBuilder.make_table_spec()`... and the one in QueryBuilder has nameshrinker logic that might also be needed here.

TallJimbo replied:
I'll look into this on [DM-43698](https://rubinobs.atlassian.net/browse/DM-43698) in order to get this branch merged faster.  I suspect  method on `ColumnSet` would work well.

dhirving commented on python/lsst/daf/butler/direct_query_driver/_driver.py:
```diff
@@ -0,0 +1,1200 @@
+        from_clause: sqlalchemy.FromClause
+        if len(dict_rows) > self._constant_rows_limit:
+            from_clause = self._exit_stack.enter_context(self.db.temporary_table(table_spec))
+            self.db.insert(from_clause, *dict_rows)
```
The Postgres documentation [recommends running ANALYZE](https://www.postgresql.org/docs/current/sql-createtable.html#SQL-CREATETABLE-TEMPORARY) on temp tables if they are going to be used in complex queries... I wonder if that is worth trying here and in materialize() above, once we get to a point where we can benchmark performance.

TallJimbo replied:
I was not aware of that recommendation, but I did think a bit about whether it would ever be useful to create indexes here, and I came to the conclusion that we probably would not be, because starting with a full table scan of the temp table and using indexes on regular tables is what we want the query optimizer to do in all cases I could imagine.  I have no idea what that says about whether ANALYZE would be helpful.

dhirving replied:
I think mainly it would let the planner know min/max values for the columns, which might affect its decisions on join order etc.  But whether the planner having that knowledge outweighs the cost of the analyze is something we'd have to benchmark probably.

dhirving commented on python/lsst/daf/butler/direct_query_driver/_driver.py:
```diff
@@ -0,0 +1,1200 @@
+                if governor in self._defaults.dataId.dimensions:
+                    result.constraint_data_id[governor] = self._defaults.dataId[governor]
+                else:
+                    raise qt.InvalidQueryError(
```
We're going to want any user-facing errors like this to be raised as a subclass of ButlerUserError in `lsst.daf.butler._exceptions`, so that they can be transferred to the client.

I can fix up InvalidQueryError to make it a ButlerUserError when I go in to do the server version of this, but if there are other exception types getting raised we'll want to keep track of that.  (The RuntimeErrors in this file are fine as-is... those are not the kind of thing that needs to go from server to client.)

TallJimbo replied:
This is one of a very small number of places where `InvalidQueryError` is raised by code I expect to move to the server; in almost all cases it's raised in client-side code (or both; it's also a model-validation error).

How about we invent a new exception for errors that come up when trying to execute a query, and replace server-side usage of `InvalidQueryError` with that?

dhirving replied:
It's fine for the same error to be raised either from client-side and server-side, and to me this makes sense as an `InvalidQueryError`.

The main distinction we need to make is between:

"this is a thing that happened because of the arguments the user specified, and the only way to fix it is for them to change their query, or add some more data to the server, or..." -> subclass of ButlerUserError, so it's re-raised on the client side.

vs

"this is a problem with the database server, an internal logic problem, an unexpected assertion failure, etc". -> not a subclass of ButlerUserError, so that the details aren't sent to the client and it triggers alerts for our team to investigate.

dhirving commented on python/lsst/daf/butler/direct_query_driver/_driver.py:
```diff
@@ -0,0 +1,1200 @@
+        self,
+        joiner: QueryJoiner,
+        resolved_search: ResolvedDatasetSearch,
+        fields: Set[str],
```
It would help clarify what is going on if instead of `str` this used the `DatasetFieldName` type alias from `query/tree/_base.py`.

(Also in `DatasetRecordStorage.make_query_joiner` and `ColumnSet.dataset_fields`.)

TallJimbo replied:
Unfortunately those aren't quite the same - there are one or two for-internal-use dataset field names that don't appear in the more user-visibile `DatasetFieldName` type alias.  It would probably make sense to define an expanded type alias for that, though.  Will do on [DM-43698](https://rubinobs.atlassian.net/browse/DM-43698).

dhirving commented on python/lsst/daf/butler/direct_butler.py:
```diff
@@ -2135,7 +2136,12 @@ def dimensions(self) -> DimensionUniverse:
+        driver = DirectQueryDriver(
+            self._registry._db, self.dimensions, self._registry._managers, self._registry.defaults
+        )
+        query = Query(driver)
```
I wonder if we also want to call `with self._caching_context` here to enable caching of collection info.

TallJimbo replied:
Yes, good idea.  Long term I think this can probably be the only way a caching context gets entered, which will hopefully let us get the caching out of the managers entirely, but entering the current caching context here is what makes sense right now.

dhirving commented on python/lsst/daf/butler/direct_query_driver/_driver.py:
```diff
@@ -0,0 +1,1200 @@
+                    else:
+                        result.append((record, self.managers.datasets.getCollectionSummary(record)))
+
+        recurse(collections)
```
I think this same operation is already implemented with somewhat fewer queries at `self.managers.datasets.fetch_summaries`.  The keys of the returned dict aren't CollectionRecord but it could easily be changed so it is -- it's operating on CollectionRecord internally.

TallJimbo replied:
I'll look into this on [DM-43698](https://rubinobs.atlassian.net/browse/DM-43698).  We might want to replace both with that recursive CTE at the same time.

dhirving commented on python/lsst/daf/butler/direct_query_driver/_query_builder.py:
```diff
@@ -0,0 +1,573 @@
+    Keys are "logical tables" - dimension element names or dataset type names.
+    """
+
+    special: dict[str, sqlalchemy.ColumnElement[Any]] = dataclasses.field(default_factory=dict)
```
I'm slightly concerned about the way assignments are made to this throughout the code... seems like there is opportunity for one chunk of code to accidentally clobber something added by another chunk.

I didn't spot any cases where that is actually occurring, though.

TallJimbo replied:
Yeah, I really struggled with how much to encapsulate `QueryBuilder` and `QueryJoiner` state throughout this ticket, and after a couple of flip-flops I made myself stop in order to keep making progress overall.  So there's room for improvement, and if you spot a way to improve it sometime while you're working nearby, feel free to give it a try; you may spot a possibility I didn't.

dhirving replied:
The only obvious idea I had was a dict that only allows one write to each key by default, so you at least have to explicitly specify clobbering.  But yeah it's not an easy problem to solve.

dhirving commented on python/lsst/daf/butler/direct_query_driver/_sql_column_visitor.py:
```diff
@@ -0,0 +1,263 @@
+        return cast(sqlalchemy.ColumnElement[Any], expression.visit(self))
+
+    def expect_timespan(self, expression: qt.ColumnExpression) -> TimespanDatabaseRepresentation:
+        return cast(TimespanDatabaseRepresentation, expression.visit(self))
```
You might consider adding instanceof asserts to these `expect_` functions instead of casting... might save some debugging time if something of an unexpected type creeps in.

dhirving commented on python/lsst/daf/butler/direct_query_driver/_query_builder.py:
```diff
@@ -0,0 +1,573 @@
+    """
+
+    def __post_init__(self) -> None:
+        assert not (self.distinct and self.group_by), "At most one of distinct and group_by can be set."
```
This check should probably be in `select()` instead of post_init... most of the assignments to distinct and group_by occur after initialization.

dhirving commented on python/lsst/daf/butler/direct_query_driver/_query_builder.py:
```diff
@@ -0,0 +1,573 @@
+    included in raw SQL results.
+    """
+
+    name_shrinker: NameShrinker | None = None
```
I wonder if it would be easier to add a `name_shrinker` property to `Database` instead of worrying about creating them and transferring records between them here.

TallJimbo replied:
I'm not sure; I'd need to look for places where we might want the `NameShrinker` without having the `Database` it came from.  If we don't have any such places (or they're easy to address some other way), this sounds like a good idea.  But one for another ticket (I'll put it on [DM-43697](https://rubinobs.atlassian.net/browse/DM-43697) for now).

dhirving reviewed (approved):
Looks good to me, just a bunch of nitpicks you can take or leave.  (Like last time GitHub is hiding some of the comments because there are too many.)

I really appreciated your extensive commenting in the hairier parts of query construction... I'm not going to claim that I fully understand all of the details but definitely closer than I was before.

Looks like it will be pretty easy to wrap the client/server version around this.

dhirving commented:
I'd like to start working on the client/server version of this early next week, while it's still fresh in my mind.  There's absolutely nothing in these review comments above the level of a petty nitpick.  So since you've got a lot on your plate at the moment I think it would be fine to merge this down as-is.

TallJimbo committed:
Reduce duplication in query result processing.

Co-authored-by: David H. Irving <david.irving@noirlab.edu>

TallJimbo committed:
Open caching context in query context.

TallJimbo committed:
Fix incorrect code comment about query transactions.

TallJimbo committed:
Use dataclass for DirectQueryDriver materialization state.

TallJimbo committed:
Drop redundant cursor close.

TallJimbo committed:
Defer guards on QueryBuilder distinct vs. group_by to make them useful.

TallJimbo committed:
Assert on rather than cast SqlColumnVisitor types.

TallJimbo committed:
Add test for DirectQueryDriver with PostgreSQL.

This has some failures that I've marked as expected until we try to fix
them again on DM-43697.
